File: tools/parity/dispatch/nacl-box.ts
// TS dispatch for the nacl-box parity module: maps the shared vector function
// names to the real `src/shared/e2ee-crypto.ts` exports so the harness compares
// the live TS reference against the Rust port.
//
// Byte buffers (Uint8Array / Vec<u8>) cross the vector boundary as plain number
// arrays so goldens stay valid JSON and compare structurally against the Rust
// side; this adapter converts at the edge and calls the real functions.
//
// `decryptBytes` needs a precomputed shared key. The Rust crate builds its
// `SharedBox` only from (secretKey, peerPublicKey) via `derive_shared_box`, so
// both halves derive the shared key the same way (`deriveSharedKey`) before
// decrypting — keeping the two adapters byte-for-byte symmetric.

import nacl from 'tweetnacl'
import { decryptBytes, deriveSharedKey } from '../../../src/shared/e2ee-crypto'

function toBytes(value: number[]): Uint8Array {
  return Uint8Array.from(value)
}

function fromBytes(bytes: Uint8Array): number[] {
  return Array.from(bytes)
}

export function dispatch(fn: string, input: unknown): unknown {
  switch (fn) {
    case 'encryptBytesWithNonce': {
      // The TS `encryptBytes` is `randomBytes(24)` + `nacl.box.after` + concat;
      // with the vector-injected nonce this reproduces its deterministic core
      // exactly (`nonce || box`), so live tweetnacl pins the Rust bundle bytes.
      const { message, nonce, ourSecretKey, peerPublicKey } = input as {
        message: number[]
        nonce: number[]
        ourSecretKey: number[]
        peerPublicKey: number[]
      }
      const sharedKey = deriveSharedKey(toBytes(ourSecretKey), toBytes(peerPublicKey))
      const nonceBytes = toBytes(nonce)
      if (nonceBytes.length !== nacl.box.nonceLength) {
        // Mirror the Rust None for a bad nonce length.
        return null
      }
      const ciphertext = nacl.box.after(toBytes(message), nonceBytes, sharedKey)
      const bundle = new Uint8Array(nonceBytes.length + ciphertext.length)
      bundle.set(nonceBytes)
      bundle.set(ciphertext, nonceBytes.length)
      return fromBytes(bundle)
    }
    case 'keyPairFromSeed': {
      const { seed } = input as { seed: number[] }
      if (seed.length !== 32) {
        return null
      }
      const pair = nacl.box.keyPair.fromSecretKey(toBytes(seed))
      return { publicKey: fromBytes(pair.publicKey), secretKey: fromBytes(pair.secretKey) }
    }
    case 'decryptBytes': {
      const { bundle, ourSecretKey, peerPublicKey } = input as {
        bundle: number[]
        ourSecretKey: number[]
        peerPublicKey: number[]
      }
      const sharedKey = deriveSharedKey(toBytes(ourSecretKey), toBytes(peerPublicKey))
      const plaintext = decryptBytes(toBytes(bundle), sharedKey)
      // decryptBytes returns null on a length/auth failure; success (even an
      // empty plaintext) is a truthy Uint8Array we surface as a number array.
      return plaintext ? fromBytes(plaintext) : null
    }
    default:
      throw new Error(`unknown function ${fn}`)
  }
}

File: rust/crates/orca-parity/src/modules/nacl_box.rs
//! Parity dispatch for `orca_crypto::nacl_box` vs `src/shared/e2ee-crypto.ts`.
//!
//! Byte buffers (`Vec<u8>` / `Uint8Array`) cross the vector boundary as plain
//! number arrays so the goldens stay valid JSON and compare structurally
//! against the TS side; this adapter converts at the edge.
//!
//! `decryptBytes` takes a precomputed shared key. This crate builds its
//! `SharedBox` only from (secret, peer-public) via `derive_shared_box`, so the
//! adapter derives it the same way the TS `deriveSharedKey` does, then decrypts.

use orca_crypto::{
    decrypt_bytes, derive_shared_box, encrypt_bytes_with_nonce, key_pair_from_seed,
    open_with_shared_key, seal_with_shared_key, shared_key_before,
};
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        // Raw-key `nacl.box.before`: the 32-byte precomputed shared key the TS
        // boundary keeps passing around (so the wasm cutover stays API-stable).
        "deriveSharedKey" => {
            let our_secret = bytes_from_json(input.get("ourSecretKey"));
            let peer_public = bytes_from_json(input.get("peerPublicKey"));
            match shared_key_before(&our_secret, &peer_public) {
                // `key` is a zeroize-on-drop wrapper; borrow the raw bytes for JSON.
                Some(key) => bytes_to_json(key.as_slice()),
                None => Value::Null,
            }
        }
        // Raw-key seal/open (`nacl.box.after` / `open.after`) — the injected
        // nonce makes the seal deterministic, pinning `nonce || box` bytes.
        "sealWithSharedKey" => {
            let shared_key = bytes_from_json(input.get("sharedKey"));
            let nonce = bytes_from_json(input.get("nonce"));
            let message = bytes_from_json(input.get("message"));
            match seal_with_shared_key(&shared_key, &nonce, &message) {
                Some(bundle) => bytes_to_json(&bundle),
                None => Value::Null,
            }
        }
        "openWithSharedKey" => {
            let shared_key = bytes_from_json(input.get("sharedKey"));
            let bundle = bytes_from_json(input.get("bundle"));
            match open_with_shared_key(&shared_key, &bundle) {
                Some(plaintext) => bytes_to_json(&plaintext),
                None => Value::Null,
            }
        }
        // Deterministic encrypt-path pin: the vector injects the nonce (the TS
        // module's only nondeterminism is `nacl.randomBytes`), so the bundle is
        // byte-exact `nonce || box` on both sides.
        "encryptBytesWithNonce" => {
            let message = bytes_from_json(input.get("message"));
            let nonce = bytes_from_json(input.get("nonce"));
            let our_secret = bytes_from_json(input.get("ourSecretKey"));
            let peer_public = bytes_from_json(input.get("peerPublicKey"));
            match derive_shared_box(&our_secret, &peer_public) {
                Some(shared) => match encrypt_bytes_with_nonce(&message, &shared, &nonce) {
                    Some(bundle) => bytes_to_json(&bundle),
                    None => Value::Null,
                },
                None => json!({ "__parity_error__": "derive_shared_box: keys must be 32 bytes" }),
            }
        }
        // X25519 keypair from a 32-byte seed (`nacl.box.keyPair.fromSecretKey`).
        "keyPairFromSeed" => {
            let seed = bytes_from_json(input.get("seed"));
            match key_pair_from_seed(&seed) {
                Some(pair) => json!({
                    "publicKey": bytes_to_json(&pair.public_key),
                    "secretKey": bytes_to_json(&pair.secret_key),
                }),
                None => Value::Null,
            }
        }
        "decryptBytes" => {
            let bundle = bytes_from_json(input.get("bundle"));
            let our_secret = bytes_from_json(input.get("ourSecretKey"));
            let peer_public = bytes_from_json(input.get("peerPublicKey"));
            match derive_shared_box(&our_secret, &peer_public) {
                // TS `decryptBytes` returns null when the box fails to open (short
                // bundle or failed Poly1305 tag); `decrypt_bytes` returns None on
                // the same paths, so the null image matches.
                Some(shared) => match decrypt_bytes(&bundle, &shared) {
                    Some(plaintext) => bytes_to_json(&plaintext),
                    None => Value::Null,
                },
                // Vectors only carry 32-byte keys; a bad length is a vector bug.
                None => json!({ "__parity_error__": "derive_shared_box: keys must be 32 bytes" }),
            }
        }
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}

fn bytes_to_json(bytes: &[u8]) -> Value {
    Value::Array(bytes.iter().map(|b| Value::from(*b)).collect())
}

fn bytes_from_json(value: Option<&Value>) -> Vec<u8> {
    value
        .and_then(Value::as_array)
        .map(|items| items.iter().filter_map(|v| v.as_u64().map(|n| n as u8)).collect())
        .unwrap_or_default()
}
